[atlas-core] allow combine for primitive types + fix Double and Date converter

mmelko opened the pull request:
fix(atlas-core DoubleConverter) double to string converter
fix(atlas-core DateConverter) hadne conversion of Long.MIN_VALUE
Fixes #266
Fixes #268
Fixes #269

mmelko committed:
fix(atlas-core combine) Allow combine for primitive type fields
fix(atlas-core DoubleConverter) double to string converter
fix(atlas-core DateConverter) hadne conversion of Long.MIN_VALUE
Fixes #266
Fixes #268
Fixes #269

mmelko commented:
@igarashitm I added also simple test but I had to create few dummy classes. It's not nice but it works. Is it ok or you would like to separate it into standalone class files.

igarashitm commented:
@mmelko thanks for the contribution! can you try the `mvn checkstyle:check` and fix a few errors?

igarashitm commented on atlas-core/src/main/java/io/atlasmap/core/BaseAtlasModule.java:
```diff
@@ -345,12 +351,11 @@ protected void processCombineField(AtlasSession session, Mapping mapping, List<F
-                // We need to support a sorted map w/ null values
-                combineValues = new HashMap<Integer, String>();
+                if (getConversionService().isPrimitive(inputField.getFieldType()))
+                    combineValues.put(inputField.getIndex(), inputField.getValue().toString());
```
Let's take even more convenient path, pick up a converter if available, which supports some of non-primitive types. Here is the example
https://github.com/atlasmap/atlasmap-runtime/blob/d4e7b4b69964f7285ccd66a4c1b14c61f3675982/atlas-core/src/main/java/io/atlasmap/core/BaseAtlasModule.java#L402-L413

igarashitm replied:
You may want to use Mockito instead. Here is the example
https://github.com/atlasmap/atlasmap-runtime/blob/d4e7b4b69964f7285ccd66a4c1b14c61f3675982/atlas-itests-parent/atlas-itests-core/src/test/java/io/atlasmap/core/issue/CamelAtlasmap14Test.java#L45-L53

mmelko replied:
You mean instead of implementation dummy classes ? (e.g https://github.com/mmelko/atlasmap-runtime/blob/combine_tweaks/atlas-core/src/test/java/io/atlasmap/core/CombineFromAtlasModuleTest.java#L78) ?

igarashitm replied:
Right, ah, the link to the specific line disappeared after your new commit :)

igarashitm commented:
And a housekeeping thing, can you add a bit description for the issue in commit comment? That helps us to find it later by `git log --grep`. Here is the example
```
fix(atlas-core combine): Allow COMBINE for non String fields

Fixes: #266
```

mmelko committed:
(atlas-core) implement conversion from byte to string
(atlas-core) remove byte check during processing COMBINE and allow combine for all types
(atlas-itests) fix jsonjson and javajson autoconversion tests

igarashitm commented on atlas-core/src/main/java/io/atlasmap/core/BaseAtlasModule.java:
```diff
@@ -342,12 +348,14 @@ protected void processCombineField(AtlasSession session, Mapping mapping, List<F
-            if (combineValues == null) {
-                // We need to support a sorted map w/ null values
-                combineValues = new HashMap<Integer, String>();
+                if (getConversionService().isPrimitive(inputField.getFieldType())) {
```
ugh I forgot to mention this, can we look for a converter whenever intputField.getFieldType() != outputField.getFieldType()? did you hit any problem? (might be another bug then...)

igarashitm replied:
i.e. not just for primitives

mmelko replied:
If converter doesn't exist exception is thrown and combine won't be performed. Would be ok to catch exception in BaseAtlasModule and log error when converter doesn't exist ?

igarashitm replied:
@mmelko +1, I think that's the best for now. Try the converter, if fails then catch the exception and fallback to object.toString(). It might be even better to put a WARN level audit if converter is not available.
